src/pyk/proof/proof.py

```diff
@@ -59,6 +61,22 @@ def dict(self) -> dict[str, Any]:
     def from_dict(cls: type[Proof], dct: Mapping[str, Any]) -> Proof:
         ...
 
+    @classmethod
+    def read_proof(cls: type[Proof], id: str, proof_dir: Path) -> Proof:
```

Comment:
Consider

```suggestion
    def read_proof(cls: type[P], id: str, proof_dir: Path) -> P:
```

where `P = TypeVar('P', Proof)`.

(This will also require `# type: ignore` for the return statement.)